expression/builtin_math_test.go

```diff
@@ -115,32 +116,65 @@ func (s *testEvaluatorSuite) TestFloor(c *C) {
 
 	sc := s.ctx.GetSessionVars().StmtCtx
 	tmpIT := sc.IgnoreTruncate
-	sc.IgnoreTruncate = true
 	defer func() {
 		sc.IgnoreTruncate = tmpIT
 	}()
 
-	for _, t := range []struct {
-		num interface{}
-		ret interface{}
-		err Checker
+	genDuration := func(h, m, s int64) types.Duration {
+		duration := time.Duration(h)*time.Hour +
+			time.Duration(m)*time.Minute +
+			time.Duration(s)*time.Second
+
+		return types.Duration{Duration: duration, Fsp: types.DefaultFsp}
+	}
+
+	genTime := func(y, m, d int) types.Time {
+		return types.Time{
+			Time: types.FromDate(y, m, d, 0, 0, 0, 0),
+			Type: mysql.TypeDatetime,
+			Fsp:  types.DefaultFsp}
+	}
+
+	for _, test := range []struct {
+		arg            interface{}
+		expect         interface{}
+		isNil          bool
+		getErr         bool
+		ignoreTruncate bool
 	}{
-		{nil, nil, IsNil},
-		{int64(1), int64(1), IsNil},
-		{float64(1.23), float64(1), IsNil},
-		{float64(-1.23), float64(-2), IsNil},
-		{"1.23", float64(1), IsNil},
-		{"-1.23", float64(-2), IsNil},
-		{"-1.b23", float64(-1), IsNil},
-		{"abce", float64(0), IsNil},
+		{nil, nil, true, false, false},
+		{int64(1), int64(1), false, false, false},
+		{float64(1.23), float64(1), false, false, false},
+		{float64(-1.23), float64(-2), false, false, false},
+		{"1.23", float64(1), false, false, false},
+		{"-1.23", float64(-2), false, false, false},
+		{"-1.b23", float64(0), false, true, false},
+		{"abce", float64(0), false, false, true},
+		{"-1.b23", float64(-1), false, false, true},
+		{genDuration(12, 59, 59), float64(125959), false, false, false},
+		{genDuration(0, 12, 34), float64(1234), false, false, false},
+		{genTime(2017, 7, 19), float64(20170719000000), false, false, false},
 	} {
-		fc := funcs[ast.Floor]
-		f, err := fc.getFunction(datumsToConstants(types.MakeDatums(t.num)), s.ctx)
+		sc.IgnoreTruncate = test.ignoreTruncate
+		f, err := newFunctionForTest(s.ctx, ast.Floor, primitiveValsToConstants([]interface{}{test.arg})...)
 		c.Assert(err, IsNil)
-		v, err := f.eval(nil)
-		c.Assert(err, t.err)
-		c.Assert(v, testutil.DatumEquals, types.NewDatum(t.ret))
+
+		result, err := f.Eval(nil)
+		if test.getErr {
+			c.Assert(err, NotNil)
+		} else {
+			c.Assert(err, IsNil)
+			if test.isNil {
+				c.Assert(result.Kind(), Equals, types.KindNull)
+			} else {
+				c.Assert(result, testutil.DatumEquals, types.NewDatum(test.expect))
+			}
+		}
 	}
+
+	f, err := funcs[ast.Floor].getFunction([]Expression{Zero}, s.ctx)
```

Comment:
I have added zero and float64 cases for signature check. Please take a look. Thanks